Bring back progressive loading

tonyd256 opened the pull request:
Loading the all the frames of a GIF at once into memory helped cut down
on the memory footprint because we could eliminate the need to hold onto
the source image. However, we see this break down when there are too
many frames. The "almost_nailed_it.gif" GIF has 545 frames and would
crash the app around 130 loaded. This brings back progressive loading
with a max frame count of 50 to prevent this issue.

kaishin commented on Source/AnimatableImageView.swift:
```diff
@@ -7,6 +7,10 @@ public class AnimatableImageView: UIImageView, Animatable {
   var animator: Animator?
 
+  deinit {
+    println("deinit animatable view")
```
This can go away.

kaishin commented on Source/Animator.swift:
```diff
@@ -34,43 +38,46 @@ class Animator: NSObject {
     pauseAnimation()
   }
 
+  deinit {
```
This too.

tonyd256 replied:
I'm currently debugging this ... it's not fully working ... I found this out after making the PR ... sorry ... we can close this if that makes sense

kaishin commented on Source/Animator.swift:
```diff
@@ -34,43 +38,46 @@ class Animator: NSObject {
-      let animatedFrame = AnimatedFrame(image: scaledImage, duration: frameDuration)
+  private func prepareFrames() {
+    numberOfFrames = Int(CGImageSourceGetCount(imageSource))
+    let framesToProcess = numberOfFrames > maxNumberOfFrames ? maxNumberOfFrames : numberOfFrames
```
`max(numberOfFrames, maxNumberOfFrames)`?

tonyd256 commented:
@kaishin This is ready for review. There are significant changes here. I started at this because the `Animator` nor the `AnimatableImageView` were `deinit`ing. There definitely was a retain cycle with `displayLink`, but there could have been a retain cycle with the delegate pattern they were using too. I removed the delegate pattern and just inject the `Animator` dependencies. I also moved the `displayLink` to the ImageView class so not the `Animator

kaishin commented on Source/AnimatableImageView.swift:
```diff
@@ -1,15 +1,19 @@
+  private lazy var displayLink: CADisplayLink = CADisplayLink(target: self, selector: Selector("updateFrame"))
+
+  /// The size of the frame cache.
+  public var frameCacheSize = 50
```
Cache is when you load something and persist it for further reuse. I think we are talking about `preload` here; when you load something ahead of its use. How do you feel abut `framePreloadCount`?

kaishin commented on Source/Animator.swift:
```diff
@@ -78,41 +79,29 @@ class Animator: NSObject {
-  func resumeAnimation() {
-    if totalDuration > 0 {
-      displayLink.paused = false
+      // load the next needed frame for progressive loading
```
"Loads the next..."

kaishin commented on Source/AnimatableImageView.swift:
```diff
@@ -50,14 +56,33 @@ public class AnimatableImageView: UIImageView, Animatable {
+    displayLink.invalidate()
+  }
+
+  /// Attaches the dsiplay link.
```
"display link" This was introduced before this PR but let's correct it anyway.

kaishin commented on Source/AnimatableImageView.swift:
```diff
@@ -50,14 +56,33 @@ public class AnimatableImageView: UIImageView, Animatable {
+
+  /// Attaches the dsiplay link.
+  private func attachDisplayLink() {
+    displayLink.addToRunLoop(NSRunLoop.mainRunLoop(), forMode: NSRunLoopCommonModes)
```
Can we get rid of `NSRunLoop.` here?

kaishin commented:
@tonyd256 If this takes care of the retain cycle, which I suspected but never got to look into, then I'm totally fine with it. As far as the API goes, this is a much lesser change than the previous one.

I'm not totally happy with my awesome Mugen gif going away, but I'll fix that once this is merged :wink:

tonyd256 commented:
@kaishin Awesome, I'll fix those things and merge it. Also, we could have a toggle button that switches the GIF ... this might actually be helpful to demonstrate how to clean up one gif and display another.
